packages/cardano-services/src/Program/services/ogmios.ts

```diff
@@ -0,0 +1,155 @@
+      if (prop === 'submitTx') {
+        return (args: Uint8Array) =>
+          ogmiosProvider.submitTx(args).catch(async (error) => {
+            if (error.innerError && isConnectionError(error.innerError)) {
+              const record = await dnsResolver(serviceName!);
+              logger.info(`DNS resolution for Ogmios service, resolved with record: ${JSON.stringify(record)}`);
+              await ogmiosProvider
+                .close?.()
+                .catch((error_) =>
+                  logger.warn(`Ogmios tx submit provider failed to close after DNS resolution: ${error_}`)
+                );
+              ogmiosProvider = ogmiosTxSubmitProvider({ host: record.name, port: record.port });
+              return await ogmiosProvider.submitTx(args);
+            }
+            throw error;
+          });
+      }
+      // Bind if it is a function, no intercept operations
+      if (typeof ogmiosProvider[prop as keyof TxSubmitProvider] === 'function') {
+        const method = ogmiosProvider[prop as keyof TxSubmitProvider] as any;
+        return method.bind(ogmiosProvider);
+      }
+    }
+  });
+};
+
+export const getOgmiosTxSubmitProvider = async (
+  dnsResolver: DnsResolver,
+  logger: Logger,
+  options?: CommonProgramOptions
+): Promise<TxSubmitProvider> => {
+  if (options?.ogmiosSrvServiceName)
+    return ogmiosTxSubmitProviderWithDiscovery(dnsResolver, logger, options.ogmiosSrvServiceName);
+  if (options?.ogmiosUrl) return ogmiosTxSubmitProvider(urlToConnectionConfig(options?.ogmiosUrl));
+  throw new MissingProgramOption(ServiceNames.TxSubmit, [
+    CommonOptionDescriptions.OgmiosUrl,
+    CommonOptionDescriptions.OgmiosSrvServiceName
+  ]);
+};
+
+/**
+ * Creates a extended OgmiosCardanoNode instance :
+ * - use passed srv service name in order to resolve the port
+ * - make dealing with failovers (re-resolving the port) opaque
+ * - use exponential backoff retry internally with default timeout and factor
+ * - intercept 'initialize' operation and handle connection errors on initialization
+ * - intercept 'eraSummaries' operation and handle connection errors runtime
+ * - all other operations are bind to pool object withoud modifications
+ *
+ * @returns OgmiosCardanoNode instance
+ */
+export const ogmiosCardanoNodeWithDiscovery = async (
+  dnsResolver: DnsResolver,
+  logger: Logger,
+  serviceName: string
+): Promise<OgmiosCardanoNode> => {
+  const { name, port } = await dnsResolver(serviceName!);
+  let ogmiosCardanoNode = new OgmiosCardanoNode({ host: name, port }, logger);
+
+  return new Proxy<OgmiosCardanoNode>({} as OgmiosCardanoNode, {
+    get(_, prop) {
+      if (prop === 'then') return;
+      if (prop === 'initialize') {
+        return () =>
+          ogmiosCardanoNode.initialize().catch(async (error) => {
+            if (isConnectionError(error)) {
+              ogmiosCardanoNode = await recreateOgmiosCardanoNode(serviceName, ogmiosCardanoNode, dnsResolver, logger);
```

Comment:
If you put `recreateOgmiosCardanoNode` inside `ogmiosCardanoNodeWithDiscovery` then you wouldn't need to pass these args. Just a note, this is good as-is.

Reply:
yep, I've just put it outside to reduce the lines of `ogmiosTxSubmitProviderWithDiscovery` even more.